### PR TITLE
[SPARK-29942][ML] Impl Complement Naive Bayes Classifier

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Impl Complement Naive Bayes Classifier as a `modelType` option in `NaiveBayes`

### Why are the changes needed?
1, it is a better choice for text classification: it is said in [scikit-learn](https://scikit-learn.org/stable/modules/naive_bayes.html#complement-naive-bayes) that 'CNB regularly outperforms MNB (often by a considerable margin) on text classification tasks.'
2, CNB is highly similar to existing MNB, only a small part of existing MNB need to be changed, so it is a easy win to support CNB.

### Does this PR introduce any user-facing change?
yes, a new `modelType` is supported

### How was this patch tested?
added testsuites

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/classification/NaiveBayes.scala [hunk 1]
```diff
@@ -204,14 +209,29 @@ class NaiveBayes @Since("1.5.0") (
...
     val piLogDenom = math.log(numDocuments + numLabels * lambda)
     var i = 0
-    aggregated.foreach { case (label, n, sumTermFreqs, _) =>
+    aggIter.foreach { case (label, n, sumTermFreqs, _) =>
       labelArray(i) = label
       piArray(i) = math.log(n + lambda) - piLogDenom
       val thetaLogDenom = $(modelType) match {
-        case Multinomial => math.log(sumTermFreqs.toArray.sum + numFeatures * lambda)
+        case Multinomial | Complement =>
+          math.log(sumTermFreqs.toArray.sum + numFeatures * lambda)
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/classification/NaiveBayes.scala]: The logic of MNB can be reused for `Complement`, we only need to 1:compute complement at first, and 2:make Matrx `theta=-theta` at the end.

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
friendly ping @srowen

File: python/pyspark/ml/classification.py [hunk 2]
```diff
@@ -1909,6 +1909,11 @@ class NaiveBayes(JavaProbabilisticClassifier, _NaiveBayesParams, HasThresholds,
     binary (0/1) data, it can also be used as `Bernoulli NB
     <http://nlp.stanford.edu/IR-book/html/htmledition/the-bernoulli-model-1.html>`_.
     The input feature values for Multinomial NB and Bernoulli NB must be nonnegative.
+    Since 3.0.0, it supports Complement NB which is an adaptation of the Multinomial NB.
+    Specifically, Complement NB uses statistics from the complement of each class to compute
+    the model's coefficients. The inventors of Complement NB show empirically that the parameter
+    estimates for CNB are more stable than those for Multinomial NB. Like Multinomial NB, the
+    input feature values for Complement NB must be nonnegative.
```
Comment (srowen) [inline python/pyspark/ml/classification.py]: Does this need any additional doc changes in Pyspark to show the new type that's allowed, or a doctest example?

**Review by srowen — COMMENTED** [review]
Looks reasonable.

File: python/pyspark/ml/classification.py (hunk 2 above)
Comment (zhengruifeng) [reply python/pyspark/ml/classification.py]: @srowen It seem that we do not need to make other doc changes. The getter and setter has no doc about supported options, I guess adding a doctest is enough.

**Review by srowen — APPROVED** [review]

**Comment by zhengruifeng** [comment]
Merged to master, thanks @srowen for reviewing!

**Pull request closed without merging**